### PR TITLE
Add a-ball handwired trackball based on ADSN9800

### PR DESCRIPTION
**Pull request by brickbots**
It's a trackball running QMK, with no buttons at all!  Designed to be used as an adjunct to another keyboard to provide pointing capabilities, while the keyboard itself provides mouse button functionality.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by brickbots** [comment]
> This should be refactored to use https://docs.qmk.fm/#/spi_driver

Thank you @zvecr for pointing out the QMK driver here.  I've refactored the code to use it 👍

File: keyboards/handwired/aball/adns.h
Comment (zvecr) [inline keyboards/handwired/aball/adns.h:1]:
```suggestion
#pragma once

```

File: keyboards/handwired/aball/keymaps/default/keymap.c
Comment (zvecr) [inline keyboards/handwired/aball/keymaps/default/keymap.c:19]:
```suggestion

```

File: keyboards/handwired/aball/keymaps/default/keymap.c
Comment (zvecr) [inline keyboards/handwired/aball/keymaps/default/keymap.c:41]:
```suggestion

```

File: keyboards/handwired/aball/adns.c
Comment (zvecr) [inline keyboards/handwired/aball/adns.c:17]:
```suggestion
```

File: keyboards/handwired/aball/rules.mk
Comment (zvecr) [inline keyboards/handwired/aball/rules.mk:36]:
```suggestion
SRC += adns.c spi_master.c
```

**Comment by brickbots** [comment]
Thank you @zvecr for your patient guidance and suggestions!  I've actioned all of your suggestions and pushed this branch again. Cheers!

File: keyboards/handwired/aball/adns9800_srom_A6.h
Comment (fauxpark) [inline keyboards/handwired/aball/adns9800_srom_A6.h:3]:
```suggestion
#include "progmem.h"
```

File: keyboards/handwired/aball/adns.c
Comment (fauxpark) [inline keyboards/handwired/aball/adns.c:144]:
```suggestion
    dprint("STARTING INTI\n");
```

File: keyboards/handwired/aball/adns.c
Comment (fauxpark) [inline keyboards/handwired/aball/adns.c:18]:
```suggestion
#include "debug.h"
```

File: keyboards/handwired/aball/adns.c
Comment (fauxpark) [inline keyboards/handwired/aball/adns.c:211]:
```suggestion
    dprint("INIT ENDED\n");
```

**Comment by brickbots** [comment]
Hi @fauxpark . Thank you for your suggestions to use the newer debug features rather than print.  Much appreciated! :+1:

**Comment by tzarc** [comment]
My only reservation with merging this is the firmware -- there's no license or comments as to where it came from.

**Comment by brickbots** [comment]
Thanks @tzarc for the review.  I've added the best information I could find about this, it seems to have been sent to the original author, MrJohnK, directly by Avago as byte code that he formatted as a c byte array for arduino use.   I hope this additional info is sufficient to help people track it back to the source.

Also, apologies for the sloppy merge there, I switched computers and had to remerge a few of the previous print->debug suggestions.  Sorry for the mess!

File: keyboards/handwired/aball/rules.mk
Comment (drashna) [inline keyboards/handwired/aball/rules.mk:36]: Since it may be sensitive to timing, I would rather see SPI master added like this: 

```suggestion
SRC += adns.c 
QUANTUM_LIB_SRC =+ spi_master.c
```

This should prevent any issues if LTO is enabled.

**Comment by brickbots** [comment]
Thank you @drashna ! I'll have to dig into the functional change here to really understand it, but I very much appreciate your suggestion about the import directives!

**Comment by brickbots** [comment]
Hello Reviewers!  Sorry if I have lost the plot here a bit.  I don't see any pending changes, but I could have missed something along the way.  Is there something holding up this merge request that I might be able to address?

File: keyboards/handwired/aball/adns.c
Comment (drashna) [inline keyboards/handwired/aball/adns.c:262]: Just a heads up, without any check here about changes in the report, this will keep the system awake.

This is an issue I ran into with the pimironi trackball module.  
If you're polling, it will cause this, but if you're using interrupts, it shouldn't be an issue.

Comment (drashna) [reply keyboards/handwired/aball/adns.c:262]: Also, the other issue here is that there is no way to customize the behavior (such as holding down a switch to switch to a "scrolling" mode, for instance)

**Review by drashna — APPROVED** [review]

File: keyboards/handwired/aball/adns.c
Comment (brickbots) [reply keyboards/handwired/aball/adns.c]: Thank you @drashna!   The heads up about keeping the machine awake makes sense as it keeps getting mouse reports. I've wrapped this so that it only occurs on change.

As for the customization... This is a bit of a strange beast, it's a trackball with NO buttons.. just a ball.  I think you have a Kyria, so you may appreciate my inspiration here:
https://github.com/brickbots/aball

Comment (drashna) [reply keyboards/handwired/aball/adns.c]: Welcome! Personally I just check if any change has occurred in the report, and not send if no change has occurred. Either way works, I think.

And that's nice! (And yup, I have a kyria)

**Comment by brickbots** [comment]
Hi reviewers!

Thank you for all the feedback and guidance so far.  I'm keen to get this merged if possible, or closed if not.  Is the semi-mysterious ADSN firmware a deal-breaker here?

**Comment by drashna** [comment]
the ADSN firmware thing should be fine.

**Review by drashna — APPROVED** [review]

**Comment by brickbots** [comment]
Hi @fauxpark and @zvecr !

I appreciate both of your time and help so far.  If I have missed a comment or change request somewhere along the line I apologize, but I'm not sure what needs to be done to get this PR merged.  Would it be please be possible to get more notes, or an approval, on this pull request?

Thanks!

File: keyboards/handwired/aball/adns.c
Comment (drashna) [inline keyboards/handwired/aball/adns.c:1]: This can be removed.

File: keyboards/handwired/aball/adns.h
Comment (drashna) [inline keyboards/handwired/aball/adns.h:1]: Also, can be removed.

File: keyboards/handwired/aball/rules.mk
Comment (drashna) [inline keyboards/handwired/aball/rules.mk:11]:
```suggestion
```

File: keyboards/handwired/aball/rules.mk
Comment (drashna) [inline keyboards/handwired/aball/rules.mk:33]:
```suggestion
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output

```

File: keyboards/handwired/aball/rules.mk
Comment (drashna) [inline keyboards/handwired/aball/rules.mk:37]:
```suggestion
POINTING_DEVICE_DRIVER = adns9800
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by brickbots** [comment]
Thanks @drashna for the suggestions to bring this PR up to snuff with the recent develop branch changes.  It's exciting to see first class support for the ADNS9800 sensor!

**Comment by drashna** [comment]
> Thanks @drashna for the suggestions to bring this PR up to snuff with the recent develop branch changes. It's exciting to see first class support for the ADNS9800 sensor!

Welcome!  And honestly, I'm happy to add first class support for as many sensors as I can. 

Also, it may be worth adding some of the defines from: 
https://github.com/qmk/qmk_firmware/pull/15304

I've been using those code changes, and they're fantastic.

File: keyboards/handwired/aball/rules.mk
Comment (fauxpark) [inline keyboards/handwired/aball/rules.mk:18]:
```suggestion
BOOTMAGIC_ENABLE = no       # Enable Bootmagic Lite
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = no        # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
NKRO_ENABLE = no            # Enable N-Key Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
```

File: keyboards/handwired/aball/config.h
Comment (fauxpark) [inline keyboards/handwired/aball/config.h:28]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]
Missing info.json.

**Comment by brickbots** [comment]
Thanks @fauxpark for the feedback on the feature description changes and the reminder about info.json!  I appreciate it 👍

**Review by drashna — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

File: keyboards/handwired/aball/info.json
Comment (fauxpark) [inline keyboards/handwired/aball/info.json:9]: Missing layouts.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/handwired/aball/info.json
Comment (brickbots) [reply keyboards/handwired/aball/info.json:9]: Yes...indeed.... but it has no buttons :-)  I can put a null there to indicate so?

Comment (drashna) [reply keyboards/handwired/aball/info.json:9]: Not really, but take a look at the ploopyco/trackball_nano to see how I handled the lack of any keys.

Comment (fauxpark) [reply keyboards/handwired/aball/info.json:16]: You have a `LAYOUT` macro with two buttons in aball.h.

Comment (brickbots) [reply keyboards/handwired/aball/info.json:11]: Aha, yes, thank you @fauxpark!  When I set this up I was not sure how to handle a matrix-less device and get it to compile.  I've now emulated what was done with the Ploopy nano and I hope it's a lot clearer and will pass the linter now as well.  Thanks for your patience!

**Review by noroadsleft — APPROVED** [review]
Looks good to me.

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**
